### PR TITLE
Skip ComputeResidualThread when there is no kernel object

### LINKED ISSUE #23894: Skip ComputeResidualThread when there is no kernel object
## Reason
Some applications may want to optimize memory usage and performance by not adding kernel objects and doing residual evaluation manually. If there is no kernel object, `ComputeResidualThread` does not have to be executed.

## Design
Add a condition that checks the existence of kernel objects before executing `ComputeResidualThread` in `NonlinearSystemBase::computeResidualInternal`.

## Impact
No impact on the interface. Performance improvement for applications that do not add kernel objects.

### PR DESCRIPTION
**Pull request by NamjaeChoi**
Closes #23894.

Related Griffin PR : https://github.inl.gov/ncrc/griffin/pull/1129

Tag @YaqiWang @zachmprince

### CODE REVIEW

File: framework/src/systems/NonlinearSystemBase.C
Comment (YaqiWang) [inline framework/src/systems/NonlinearSystemBase.C:1564]: Maybe move that check inside `NonlinearThread::hasObjects`? I feel it is safer in this way because it sees directly all the objects. We can still build this `ComputeResidualThread`, but will call parallel_reduce conditionally.

Comment (loganharbour) [reply framework/src/systems/NonlinearSystemBase.C:1563]: Yeah, at the minimum. I'm still a bit uncomfortable about this. What's so bad about looping through the elements to realize we have no objects?

**Comment by lindsayad** [comment]
I don't understand why I didn't get triggered for review on this. Wtf CODEOWNERS?

File: framework/src/systems/NonlinearSystemBase.C
Comment (zachmprince) [reply framework/src/systems/NonlinearSystemBase.C:1563]: I don't understand why we need to loop through the mesh... This is at least obvious about what we are skipping. We have similar logic below for nodal kernels.

Comment (roystgnr) [reply framework/src/systems/NonlinearSystemBase.C:1563]: > What's so bad about looping through the elements to realize we have no objects?

`O(N_elem)` is asymptotically worse than `O(0)`?  Even on DistributedMesh the element loop is something like `O(N_elem/N_proc * log(N_elem/N_proc))`; we don't have to predicate our way past a zillion non-local elements but we do have to iterate through an ordered map (or a more complicated data structure that has a better constant but the same asymptotics) under the hood.

Comment (loganharbour) [reply framework/src/systems/NonlinearSystemBase.C:1563]: > I don't understand why we need to loop through the mesh... This is at least obvious about what we are skipping. We have similar logic below for nodal kernels.

Because it is not immediately clear if we add capability to the threaded loop object to execute more objects, but then forget to update this change.

I'm not completely against this -  but this logic should at the very least be local to the threaded loop object, not a System method.

Comment (loganharbour) [reply framework/src/systems/NonlinearSystemBase.C:1563]: I like the idea of adding an early exit to the threaded object if there is nothing to do.

Comment (NamjaeChoi) [reply framework/src/systems/NonlinearSystemBase.C:1563]: So, the approach what you guys agree is to implement a `hasActiveObjects` function inside `NonlinearThread` and use that to skip the calculation?

Comment (NamjaeChoi) [reply framework/src/systems/NonlinearSystemBase.C:1563]: Or add a check in `onElement` on the existence of objects like other routines like `onInternalSide` and do an empty loop?

Comment (NamjaeChoi) [reply framework/src/systems/NonlinearSystemBase.C:1563]: Once MOOSE maintainers confirm, I will push it again.

File: framework/src/loops/NonlinearThread.C
Comment (lindsayad) [inline framework/src/loops/NonlinearThread.C:59]:
```suggestion
    ThreadedElementLoop<ConstElemRange>::operator()(range, bypass_threading);
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/include/loops/NonlinearThread.h
Comment (lindsayad) [inline framework/include/loops/NonlinearThread.h:112]:
```suggestion
  /// Whether there are any active residual objects; otherwise we will do an early return
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
Sorry for not suggesting this with the last one.

@GiudGiud do you see any conceptual clash with the `_has_active_objects` logic and the refactoring you did in your debug objects PR?

**Comment by GiudGiud** [comment]
I dont think so.
I also query "whether I have an active object" but on each block.

File: framework/src/loops/NonlinearThread.C
Comment (GiudGiud) [inline framework/src/loops/NonlinearThread.C:59]: are you going to be ok skipping the jacobian loop?

It s not a NonLinearThread  inherited class for now, but it will be soon

Comment (YaqiWang) [reply framework/src/loops/NonlinearThread.C:59]: I was checking that. Looks like it is good. `ComputeJacobianThread` has only these four warehouses as well. And the skipping should be logically correct.

**Review by lindsayad — APPROVED** [review]

File: framework/src/loops/NonlinearThread.C
Comment (GiudGiud) [reply framework/src/loops/NonlinearThread.C:59]: same warehouses right now between Jacobian and Residual. But yes

**Comment by lindsayad** [comment]
I'll merge if/when the conversation is resolved

**Comment by GiudGiud** [comment]
done. I dont think this is an issue here. Happy to know about this though

**Pull request merged**
